images/cfw/opt/x1plus/lib/python/x1plus/services/x1plusd/mcproto.py

```diff
@@ -0,0 +1,353 @@
+    (1, 4): "sync_timestamp",
+    (1, 6): "mcu_upgrade",
+    (1, 8): "mcu_hms",
+    (1, 9): "factory_reset",
+    (2, 5): "gcode_execute_state",
+    (2, 6): MC_gcode_request, # no official name for this
+    (2, 9): MC_mcu_display_message,
+    (2, 10): "vosync",
+    (2, 11): "gcode_ctx",
+    (2, 12): "mc_state",
+    (2, 15): "link_ams_report",
+    (2, 17): "ack_tray_info",
+    (2, 22): "gcode_line_handle",
+    (2, 23): "ams_mapping",
+    (2, 24): "ams_tray_info_write_ack",
+    (2, 25): "ams_user_settings",
+    (2, 27): "hw_info_voltage",
+    (2, 28): "link_ams_tray_consumption_ack",
+    (2, 29): "pack_get_part_info_ack",
+    (2, 34): "extrusion_result_update",
+    (2, 36): "fila_ams_get",
+    (2, 37): "mc_get_skipped_obj_list",
+    (3, 1): MC_M971,
+    (3, 2): "M972",
+      # has to do with camera clarity, "calculate intrinsic params"
+      # payload_len < 6
+      #   0, 1: seq
+      #   2: "on"
+      #   3: "data_num"
+    (3, 5): "M963",
+      # during handeye calibration sequence, syntax "M963 S1"
+      # "calculate extrinsic params"
+    (3, 7): "M969",
+      # syntax "M969 S1 N3 A2000" or "M969 S0 ; turn off scanning"
+      # "calculate extruder bandwidth"
+    (3, 6): "M965_b",
+    (3, 9): "M967",
+    (3, 11): "M973", # "ctask set camera exposure"
+      # M973 S6 P0; use auto exposure for horizontal laser by xcam
+      # M973 S6 P1; use auto exposure for vertical laser by xcam
+      # M973 S3 P14
+      # M973 S3 P1 ; camera start stream
+      # M973 S1
+      # M973 S2 P0
+      # M973 S4 ; turn off scanner
+    (3, 14): "M965", # "calculate heatbed height measurement" or "calculate baseline heatbed measurement"
+    (3, 49): MC_M976,
+    (3, 50): "M977",
+      # "detection single layer registration"
+    (3, 51): "M978",
+      # "calculate continuous layer detection"
+    (3, 52): "M981",
+      # "calculate spaghetti detection"
+      # M981 S1 P20000 ;open spaghetti detector
+    (3, 53): "M991",
+      # "layer change event"
+    (3, 81): "M987",
+    (3, 82): "SENSORCHECK",
+      # "set sensor detection"
+    (4, 1): "set_module_sn",
+    (4, 2): "get_module_sn",
+    (4, 3): "inject_module_key",
+    (4, 4): "get_inject_status",
+    (4, 5): "mcu_reboot",
+    (4, 6): "send_to_amt_core",
+    (4, 7): "set_module_lifecycle",
+    (4, 8): "get_module_lifecycle",
+    (4, 10): "inject_productcode",
+    (4, 11): "get_productcode",
+}
+
+###
+
+class MCProtoParser():
+    """
+    MC protocol parser service.
+    
+    MCProtoParser handles DBus signals sent over by forward, as glued in by
+    forward_shim.  It parses all outbound messages to the MC, as well as
+    inbound messages, and parcels out things that it needs to care about --
+    for instance, it does its own parsing on all outbound Gcode (to find
+    x1plus metadata), and listens for M976s to trigger X1Plus-specific
+    behavior.
+    
+    This comes at a cost -- about 15% of a core of CPU (ugh!).  But luckily
```

Comment:
Yeah, this is CPU used by `mcproto.py`.  Fascinatingly, we have plenty of CPU to go around.  But yes, it would be nice to have some kind of data on this sooner or later.